src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs

```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Threading.Tasks;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// Creates a zip archive with all the .csproj and .targets encountered during the build.
+    /// The internal .zip file structure matches closely the layout of the original sources on disk.
+    /// The .zip file can be used to correlate the file names and positions in the build log file with the
+    /// actual sources.
+    /// </summary>
+    internal class ProjectImportsCollector
+    {
+        private FileStream _fileStream;
+        private ZipArchive _zipArchive;
+
+        public string ArchiveFilePath { get; set; }
+
+        /// <summary>
+        /// Avoid visiting each file more than once.
+        /// </summary>
+        private readonly HashSet<string> _processedFiles = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        // this will form a chain of file write tasks, running sequentially on a background thread
+        private Task _currentTask = Task.CompletedTask;
+
+        public ProjectImportsCollector(string logFilePath, string sourcesArchiveExtension = ".projectimports.zip")
+        {
+            ArchiveFilePath = Path.ChangeExtension(logFilePath, sourcesArchiveExtension);
+
+            try
+            {
+                _fileStream = new FileStream(ArchiveFilePath, FileMode.Create, FileAccess.ReadWrite, FileShare.Delete);
+                _zipArchive = new ZipArchive(_fileStream, ZipArchiveMode.Create);
+            }
+            catch
+            {
+                // For some reason we weren't able to create a file for the archive.
+                // Disable the file collector.
+                _fileStream = null;
+                _zipArchive = null;
+            }
+        }
+
+        public void AddFile(string filePath)
+        {
+            if (filePath == null || _fileStream == null)
+            {
+                return;
```

Comment:
As far as I can see, it is valid and can be hit.

Let's first deal with filePath. AddFile() is called from two places and in one of them we pass ProjectStartedEventArgs.ProjectFile, which can be null: http://source.dot.net/#Microsoft.Build.Framework/ProjectStartedEventArgs.cs,8d59ac5ef1c85c0b Unless we can prove that that is never null (which I don't think we can) we should expect null as a valid value for filePath.

_fileStream can be null if the collector was disabled due to a previous error (such as unable to write to the log file).

In non-essential functionality such as here I'd prefer to fail silently than risk crashing the build. Worst case, we don't include a file, who cares :)